### PR TITLE
Mrmiywj

### PR DESCRIPTION
**Pull request by mrmiywj**
Some problem with recursive function and closure.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1 +1,2 @@
 target/
+.idea/
```
Comment (gzm0) [inline .gitignore]: Please don't put IDE specific stuff in the project .gitignore (put it in your global instead). I know that the Scala.js project itself differs on that for historic reasons. For new projects we'd stick with that convention.

**Comment by gzm0** [comment]
Try feeding this to your compiler:

```
if (1) let x = 1 in x *2 else x * 3
```

File: src/test/scala/calc/RunTest.scala [hunk 2]
```diff
@@ -18,7 +19,7 @@ import org.scalajs.jsenv.JSConsole
  *  result.
  */
 class RunTest {
-
+  val r = Random
```
Comment (gzm0) [inline src/test/scala/calc/RunTest.scala]: Consider seeding your random to get deterministic tests (not really relevant here I know, but in general very relevant).

**Comment by sjrd** [comment]
> Some problem with recursive function

Since there is no `letrec` statement in the source language, it is not expected that recursive functions work.

File: src/main/scala/calc/Compiler.scala [hunk 3]
```diff
@@ -3,6 +3,7 @@ package calc
 import org.scalajs.core.ir
 import ir.{Trees => irt, Types => irtpe}
 import ir.Definitions._
+import scala.collection.mutable.Map
```
Comment (sjrd) [inline src/main/scala/calc/Compiler.scala]: Using mutable stuff is a code smell. It is expected that this compiler be written in a purely functional style.

File: src/test/scala/calc/RunTest.scala (hunk 2 above)
Comment (mrmiywj) [inline src/test/scala/calc/RunTest.scala]: Hmm, I don't think `if (1) let x = 1 in x*2 else x*3` is a valid statement. Because the scope of `x` is just the then phase of the conditional statement, not including else phase.

File: src/main/scala/calc/Compiler.scala [hunk 4]
```diff
@@ -74,8 +99,73 @@ object Compiler {
...
+          case "-" => irt.BinaryOp.Double_-
+          case "*" => irt.BinaryOp.Double_*
+          case "/" => irt.BinaryOp.Double_/
+        }
+        irt.BinaryOp(jsop, irt.Unbox(compileExpr(lhs),'D'), irt.Unbox(compileExpr(rhs), 'D'))
+      }
+
+      case Let(ident, value, body) => {
+        var jsV:irt.Tree = if (value.isInstanceOf[Closure]){
+          compileClosure(value.asInstanceOf[Closure], ident.name)
```
Comment (sjrd) [inline src/main/scala/calc/Compiler.scala]: Oh so you basically did interpret `let` as a `letrec`.
Could make sense, but translation into the IR will be very hard. It is not asked that you support recursion.

File: src/main/scala/calc/Compiler.scala [hunk 5]
```diff
@@ -74,8 +99,73 @@ object Compiler {
...
-      case _ =>
-        throw new Exception(
+      case BinaryOp(op, lhs, rhs) => {
+        val jsop = op match {
+          case "+" => irt.BinaryOp.Double_+
+          case "-" => irt.BinaryOp.Double_-
+          case "*" => irt.BinaryOp.Double_*
+          case "/" => irt.BinaryOp.Double_/
+        }
+        irt.BinaryOp(jsop, irt.Unbox(compileExpr(lhs),'D'), irt.Unbox(compileExpr(rhs), 'D'))
```
Comment (sjrd) [inline src/main/scala/calc/Compiler.scala]: Why do you need to unbox the arguments, here? You should be able to statically know whether they are numbers (in which case it's fine) or functions (in which case it's a compile-time type error).

File: src/test/scala/calc/RunTest.scala (hunk 2 above)
Comment (sjrd) [inline src/test/scala/calc/RunTest.scala]: I guess this was related to https://github.com/sjrd/scalajsir-calc/pull/2#issuecomment-199785139 ?

I think @gzm0's point is that, because your mutable map `idToType`, your implementation will actually _accept_ this program, whereas it should report a compile-time error, precisely because of the scoping issue.

Comment (gzm0) [inline src/test/scala/calc/RunTest.scala]: > I think @gzm0's point is that, [...]

Yes

Comment (mrmiywj) [inline src/test/scala/calc/RunTest.scala]: Sure. I got it! Since it's just a prototype, I did not care some corner cases. Just.. hmm, works on some common cases

File: src/main/scala/calc/Compiler.scala (hunk 5 above)
Comment (mrmiywj) [inline src/main/scala/calc/Compiler.scala]: Hmm.. Since I have not implemented a type checker now, I just force it to be `Double` type. I will remove it after type checking.

**Comment by mrmiywj** [comment]
Hi, 
Some changes:
1. pure functional implementation
2. Support closure and function call

Problem maybe:
No type checker now, so I just cast the binary operand's type.

File: src/main/scala/calc/Compiler.scala [hunk 6]
```diff
@@ -74,8 +99,72 @@ object Compiler {
...
+
+      case Closure(idList, body) => {
+        /*val capturePara = idList.map(s => s match{
+          case Ident(n) => irt.ParamDef(irt.Ident(n), irtpe.AnyType, false, false)
+          case _ => throw new Exception("Function def error!")
+        })*/
+        val captureSet = findCapture(tree)
+        val captureParam = captureSet.map(s => irt.ParamDef(irt.Ident(s), typeEnv(s), false, false)).toList
+        val params = idList.map(s => irt.ParamDef(irt.Ident(s.name), irtpe.AnyType, false, false))
+        irt.Closure(captureParam, params, compileExpr(body)(env, typeEnv), captureSet.map(s => env(s)).toList)
```
Comment (sjrd) [inline src/main/scala/calc/Compiler.scala]: `s => env(s)` is duplicating the rhs (right-hand-side) of `let` expressions if they are captured. You should simply use a `irt.VarRef` to the variable declared by the `let` expression.

File: src/main/scala/calc/Compiler.scala [hunk 7]
```diff
@@ -61,12 +64,34 @@ object Compiler {
...
+  }
+
+
   /** Compile an expression tree into an IR `Tree`, which is an expression
    *  that evaluates to the result of the tree.
    *
    *  This is the main method you have to implement.
    */
-  def compileExpr(tree: Tree): irt.Tree = {
+  def compileExpr(tree: Tree)(implicit env:Map[String, irt.Tree],typeEnv:Map[String, irtpe.Type]): irt.Tree = {
```
Comment (sjrd) [inline src/main/scala/calc/Compiler.scala]: Rather than using implicit parameters, I recommend define a private overload of `compileExpr` (or even with a different name). It is not good practice to use implicit parameters of "general" types (like `Map`).

File: src/main/scala/calc/Compiler.scala [hunk 8]
```diff
@@ -61,12 +64,34 @@ object Compiler {
...
+
+  def findCapture(tree: Tree): Set[String] = {
+    tree match {
+      case _:Literal => Set()
+      case BinaryOp(_, l, r) => findCapture(l) ++ findCapture(r)
+      case Let(i, v, body) => findCapture(body) ++ findCapture(v) - i.name
+      case Ident(n) => Set(n)
+      case If(cond, thenp, elsep) => findCapture(cond) ++ findCapture(thenp) ++ findCapture(elsep)
+      case Closure(idents, body) => findCapture(body) -- idents.toSet.flatMap((x:Tree) => findCapture(x))
+      case Call(f, args) => findCapture(f) ++ args.toSet.flatMap((x:Tree) => findCapture(x))
```
Comment (sjrd) [inline src/main/scala/calc/Compiler.scala]: Calling `args.toSet` is weird, because having a `Set[Tree]` is weird. `Tree`s are not meant to be compared for equality, and especially their hash code is expensive to compile (it has to walk the entire subtree). You should first `flatMap` _then_ `.toSet`, so that you never have a transient `Set[Tree]` (only `Set[String]`).

File: src/main/scala/calc/Compiler.scala [hunk 9]
```diff
@@ -61,12 +64,34 @@ object Compiler {
...
+    }
+  }
+
+  def compileClosure(tree: Closure, ident: String)(implicit env:Map[String, irt.Tree],typeEnv:Map[String, irtpe.Type]): irt.Tree ={
+    val captureSet = findCapture(tree) - ident
+    val captureParam = captureSet.map(s => irt.ParamDef(irt.Ident(s)(tree.pos), typeEnv(s), false, false)(tree.pos)).toList
+    val params = tree.params.map(s => irt.ParamDef(irt.Ident(s.name)(tree.pos), irtpe.AnyType, false, false)(tree.pos))
+    irt.Closure(captureParam, params, compileExpr(tree.body)(env, typeEnv), captureSet.map(s => env(s)).toList)(tree.pos)
```
Comment (sjrd) [inline src/main/scala/calc/Compiler.scala]: Same note about `s => env(s)` as below.

File: src/main/scala/calc/Compiler.scala [hunk 10]
```diff
@@ -61,12 +64,34 @@ object Compiler {
...
+      case If(cond, thenp, elsep) => findCapture(cond) ++ findCapture(thenp) ++ findCapture(elsep)
+      case Closure(idents, body) => findCapture(body) -- idents.toSet.flatMap((x:Tree) => findCapture(x))
+      case Call(f, args) => findCapture(f) ++ args.toSet.flatMap((x:Tree) => findCapture(x))
+    }
+  }
+
+  def compileClosure(tree: Closure, ident: String)(implicit env:Map[String, irt.Tree],typeEnv:Map[String, irtpe.Type]): irt.Tree ={
+    val captureSet = findCapture(tree) - ident
+    val captureParam = captureSet.map(s => irt.ParamDef(irt.Ident(s)(tree.pos), typeEnv(s), false, false)(tree.pos)).toList
+    val params = tree.params.map(s => irt.ParamDef(irt.Ident(s.name)(tree.pos), irtpe.AnyType, false, false)(tree.pos))
```
Comment (sjrd) [inline src/main/scala/calc/Compiler.scala]: So, here, you define parameters of type `any`. And then, _every_ time a parameter is used, you cast it down to a `double` with `Unbox(x, 'D')`. In the typical spirit of erasure for Java and Scala, it would be better to downcast them _once_, upon entry in the function.

That means parameters will need a name different than `s.name` (a generated one), since that name will be used for a local `VarDef` of type `Double`, which will store the result of `Unbox`. Then that local variable can be used in the body of the method.

Comment (sjrd) [inline src/main/scala/calc/Compiler.scala]: With this change, and even without a separate type checker, the `Unbox` that you have `BinaryOp` will become unnecessary, as will many others.

File: src/test/scala/calc/RunTest.scala [hunk 11]
```diff
@@ -38,4 +39,48 @@ class RunTest {
...
+
+    testCapture("let x=1 in x+y", Set("y"))
+    testCapture("let x=1 in let y=2 in u+v", Set("u","v"))
+    testCapture("let f=fun (x) = {let y=1 in x+y} in f (1)", Set())
+    testCapture("let f=fun (g) = {g (1)} in let g=fun (x) = {x+1} in f (g)", Set())
+    testCapture("let f=fun (g) = {g (x)} in f (y)", Set("x", "y"))
+    testCapture("fun (x) = {x + 1}", Set())
+    testCapture("let fac = fun (x) = {if (x) x*fac(x-1) else 1} in fac(5)", Set())
+    testCapture("let x=1 in let f=fun(y)={y+x} in f(1)", Set())
+    testCapture("let f=fun(y)={y+x} in f(1)", Set("x"))
```
Comment (sjrd) [inline src/test/scala/calc/RunTest.scala]: These are good tests!
Ideally, they should be in `CompilerTest`, though, as they do not run anything, but rather test the internals of the compiler.

File: src/test/scala/calc/RunTest.scala [hunk 12]
```diff
@@ -38,4 +39,48 @@ class RunTest {
...
+  }
+
+  @Test def runFunAndCall(): Unit = {
+    assertRun(2.5, "let x=1 in let f=fun(y)={y+x} in f(1.5)")
+    assertRun(2.5, "let f=fun(g)={g(1.0)} in let g=fun(x)={x+1.5} in f(g)")
```
Comment (sjrd) [inline src/test/scala/calc/RunTest.scala]: I see here that you somehow support higher-order functions.

I guess this works because your compiler basically treats the program as dynamically typed, and defers _all_ type checks at run-time. Writing a type-checker for this language with HoFs will be insanely hard, compared to the expected difficulty of this assignment. It goes well beyond what is expected here. Essentially, I believe you'll need Hindley-Milner type inference, which is Haskell-like, and not at all Java-like nor Scala-like.

Comment (mrmiywj) [inline src/test/scala/calc/RunTest.scala]: Thanks for all your advice! I will modify my code after my exam tonight(maybe 2p.m. in your zone)

**Comment by mrmiywj** [comment]
Hi,
According to your comments, I just modified my code and make it type matched as I understood. But unlikely, it does not run simple cases like `let f=fun(x)={x+1} in f(1)` well. I thought I must mistake something. Could you offer any advice? Thx:)

File: src/main/scala/calc/Compiler.scala [hunk 13]
```diff
@@ -74,8 +98,60 @@ object Compiler {
...
+          throw new Exception(s"The then and also phase in the conditional statement must bse same type")
+        }
+        irt.If(finalCond, jsThenp, jsElsep)(jsThenp.tpe)
+      }
+
+      case Closure(idList, body) => {
+        val captureList = findCapture(tree).toList
+        val captureParam = captureList.map(s => irt.ParamDef(irt.Ident(s), typeEnv(s), false, false))
+        val params = idList.map(s => irt.ParamDef(irt.Ident(s.name), irtpe.AnyType, false, false))
+        val newTypeEnv = idList.foldLeft(typeEnv) ((env:Map[String,irtpe.Type], s:Ident) => env+(s.name -> irtpe.DoubleType))
```
Comment (sjrd) [inline src/main/scala/calc/Compiler.scala]: I think you're mistake is here. You create params with `s.name` and type `any`, but you add in your `envType` that `s.name` has type `double`.

Instead, you should declare params with other (generated) names, of type `any`, and add `VarDef`s at the beginning of the method to unbox the params in local variable of name `s.name` and type `double`. Then it's ok to use `s.name` of type `double` in the environment.

**Comment by mrmiywj** [comment]
The calculator has been done! Need your advice for next step!

File: src/main/scala/calc/Compiler.scala [hunk 14]
```diff
@@ -61,21 +65,89 @@ object Compiler {
...
+          case "*" => irt.BinaryOp.Double_*
+          case "/" => irt.BinaryOp.Double_/
+        }
+        irt.BinaryOp(jsop, compileExpr(lhs)(env, typeEnv), compileExpr(rhs)(env, typeEnv))
+
+      case Let(ident, value, body) =>
+        val jsV = compileExpr(value)(env, typeEnv)
+        val jsIdent = irt.Ident(ident.name)
+        val jsId = irt.VarDef(jsIdent, jsV.tpe, mutable = false, jsV)
+        irt.Block(List(jsId, compileExpr(body)(env + (ident.name -> jsV), typeEnv + (ident.name -> jsV.tpe))))
```
Comment (sjrd) [inline src/main/scala/calc/Compiler.scala]: It is a code smell that you still have `env` at all. `typeEnv` is necessary, of course, but `env` should never be needed, since the value tree should only ever appear as the rhs of the `VarDef`, but in the body you should only use the newly declared variable.

File: src/main/scala/calc/Compiler.scala [hunk 15]
```diff
@@ -61,21 +65,89 @@ object Compiler {
...
+
+      case Ident(n) =>
+        try {
+          irt.VarRef(irt.Ident(n))(typeEnv(n))
+        }catch{
+          case e:Exception => throw new Exception("No such identity!")
```
Comment (sjrd) [inline src/main/scala/calc/Compiler.scala]: identity -> identifier ;)

File: src/main/scala/calc/TypeChecker.scala [hunk 16]
```diff
@@ -0,0 +1,34 @@
...
+
+case object Double extends Type
+
+case class Fun(arg: Type, body: Type) extends Type
+
+/**
+  * Created by ivan on 16/3/23.
+  * A type checker and infer about types
+  */
+object TypeChecker {
```
Comment (sjrd) [inline src/main/scala/calc/TypeChecker.scala]: It seems this is unused.

**Comment by sjrd** [comment]
AFAICT, you're still missing the standard math functions such as `sin`, `sqrt`, etc.

**Comment by mrmiywj** [comment]
Complete type checker.

TODOs:
- Add common math functions

File: src/main/scala/calc/Compiler.scala [hunk 17]
```diff
@@ -61,21 +65,93 @@ object Compiler {
...
+  }
+
+  var seed = 1
+  def generateName():String = {
+    seed += 1
+    "generated_"+seed.toString
+  }
+
+  def compileExpr(tree: Tree):irt.Tree ={
+    val treeType = typeInfer(tree)
```
Comment (sjrd) [inline src/main/scala/calc/Compiler.scala]: `treeType` is unused.

File: src/main/scala/calc/Compiler.scala [hunk 18]
```diff
@@ -61,21 +65,93 @@ object Compiler {
...
+          irt.VarRef(irt.Ident(n))(typeEnv(n))
+        }catch{
+          case e:Exception => throw new Exception("No such identitier!")
+        }
+
+      case If(cond, thenp, elsep) =>
+        val jsThenp = compileExpr(thenp, typeEnv)
+        val jsElsep = compileExpr(elsep, typeEnv)
+        val jsCond = compileExpr(cond, typeEnv)
+        val finalCond = irt.BinaryOp(irt.BinaryOp.!==, irt.Unbox(jsCond,'D'), irt.DoubleLiteral(0.0))
```
Comment (sjrd) [inline src/main/scala/calc/Compiler.scala]: There should be no need for `Unbox` here, as the type checker asserted that `jsCond` is a double.

File: src/main/scala/calc/TypeChecker.scala [hunk 19]
```diff
@@ -0,0 +1,52 @@
...
+case object DoubleType extends TreeType
+
+case class FunType(numOfArgs: Int) extends TreeType
+
+/**
+  * Created by ivan on 16/3/23.
+  * A type checker and infer about types
+  */
+object TypeChecker {
+  def typeInfer(tree: Tree, typeEnv: Map[String, TreeType] = Map()): TreeType ={
```
Comment (sjrd) [inline src/main/scala/calc/TypeChecker.scala]: I suggest you call this method `typeCheck` instead of `typeInfer`, since it essentially performs type checking.

File: src/main/scala/calc/TypeChecker.scala [hunk 20]
```diff
@@ -0,0 +1,52 @@
...
+    tree match {
+      case Literal(value) => DoubleType
+      case Ident(name) => typeEnv(name)
```
Comment (sjrd) [inline src/main/scala/calc/TypeChecker.scala]: This will throw a not-very-nice exception for the developer if the identifier is not in the scope.

File: src/main/scala/calc/TypeChecker.scala [hunk 21]
```diff
@@ -0,0 +1,52 @@
...
+      case BinaryOp(op, lhs, rhs) => {
+        val ltype = typeInfer(lhs, typeEnv)
+        val rtype = typeInfer(rhs, typeEnv)
+        if (ltype == rtype) DoubleType
```
Comment (sjrd) [inline src/main/scala/calc/TypeChecker.scala]: This is not strict enough. Both operands must be doubles. But this test would pass if both operands are functions.

File: src/main/scala/calc/TypeChecker.scala [hunk 22]
```diff
@@ -0,0 +1,52 @@
...
+        else throw new Exception("Binary operands does not match!")
+      }
+      case Let(n, value, body) => {
+        val valueType = typeInfer(value, typeEnv)
+        typeInfer(body, typeEnv + (n.name -> valueType))
+      }
+      case Closure(params, body) =>
+        FunType(params.length)
```
Comment (sjrd) [inline src/main/scala/calc/TypeChecker.scala:51]: You're not typechecking the `body`, here.

File: src/main/scala/calc/TypeChecker.scala [hunk 23]
```diff
@@ -0,0 +1,52 @@
...
+      case Call(fun, args) => {
+        val funType = typeInfer(fun, typeEnv)
+        if (args.length == funType.asInstanceOf[FunType].numOfArgs) DoubleType
```
Comment (sjrd) [inline src/main/scala/calc/TypeChecker.scala:57]: If the type of `fun` is not a function, this will cause an ugly `ClassCastException`.

Comment (sjrd) [inline src/main/scala/calc/TypeChecker.scala:57]: You're also not typechecking the `args`. They must all be doubles.

**Comment by mrmiywj** [comment]
Changes:
- Type checker errors
- Add common math functions
- Modify the parser to support direct function call
- Coding style

File: src/main/scala/calc/Compiler.scala [hunk 24]
```diff
@@ -61,21 +62,113 @@ object Compiler {
...
+      case Closure(idents, body) => findCapture(body) -- idents.flatMap((x:Tree) => findCapture(x)).toSet
+      case Call(f, args) => findCapture(f) ++ args.flatMap((x:Tree) => findCapture(x)).toSet
+    }
+  }
+
+
+  def compileExpr(tree: Tree):irt.Tree ={
+    implicit val pos = tree.pos
+    typeCheck(tree)
+    implicit val mathModule = irt.LoadModule(irtpe.ClassType("jl_Math$"))
```
Comment (sjrd) [inline src/main/scala/calc/Compiler.scala]: It would be much better to create this tree where you need it, when you need it. For starters, you wouldn't have to thread it around the entire code. But also it would have the appropriate `pos` at each place where it is needed.

Also, it makes more sense, independently of this assignment. In a real compiler, you could not afford passing around such hard-coded trees for all sorts of stuff you might need everywhere.

File: src/main/scala/calc/Compiler.scala [hunk 25]
```diff
@@ -61,21 +62,113 @@ object Compiler {
...
+        try {
+          val jsFunction = compileExpr(f, typeEnv)
+          irt.Unbox(irt.JSFunctionApply(jsFunction,args.map(s => compileExpr(s, typeEnv))), 'D')
+        }catch{
+          case e:Exception=> {
+            try {
+              val funcName = f.asInstanceOf[Ident].name
+              val identName = funcName + "__D" * functionTypes(funcName) + "__D"
+              irt.Apply(mathModule, irt.Ident(identName),
+                args.map(s => compileExpr(s, typeEnv)))(irtpe.DoubleType)
```
Comment (sjrd) [inline src/main/scala/calc/Compiler.scala]: Basically, here you could just write `irt.LoadModule(irtpe.ClassType("jl_Math$"))` insted of `mathModule`, and be done with it.

File: src/main/scala/calc/Compiler.scala [hunk 26]
```diff
@@ -61,21 +62,113 @@ object Compiler {
...
+            irt.Unbox(irt.VarRef(irt.Ident("$$"+s.name))(irtpe.AnyType), 'D')))
+        val bl = irt.Block(unboxParams ++ List(compileExpr(body, newTypeEnv)))
+        irt.Closure(captureParam, params, bl, captureList.map(s => irt.VarRef(irt.Ident(s))(typeEnv(s))))
+
+      case Call(f, args) =>
+        try {
+          val jsFunction = compileExpr(f, typeEnv)
+          irt.Unbox(irt.JSFunctionApply(jsFunction,args.map(s => compileExpr(s, typeEnv))), 'D')
+        }catch{
+          case e:Exception=> {
```
Comment (sjrd) [inline src/main/scala/calc/Compiler.scala]: This is a code smell. You're dealing with a completely normal case by throwing an exception and catching it. Worse, you're catching _any_ type of `Exception`.

It would be much nicer to preemptively _test_ whether `f` is a math function (and not a function in the lexical scope; which potentially shadows a standard function).

To do this, you might have to special-case a `case Call(Ident(name), args) if !typeEnv.contains(name) =>` or something like that.

This of course highlights the fact that you can only deal with direct calls to math functions (and not, e.g., `let f = sin in f(1)`), but that's ok. If you want an additional challenge, you could try to address that, but that's not necessary.

File: src/main/scala/calc/TypeChecker.scala [hunk 27]
```diff
@@ -0,0 +1,72 @@
...
+                         "max" -> FunType(2),
+                         "min" -> FunType(2))
+  def typeCheck(tree: Tree, typeEnv: Map[String, TreeType] = Map()): TreeType ={
+    tree match {
+      case Literal(value) => DoubleType
+      case Ident(name) =>
+        try {
+          typeEnv(name)
+        }catch{
+          case e:Exception => if (mathFunctions.contains(name))
```
Comment (sjrd) [inline src/main/scala/calc/TypeChecker.scala]: Same here with the catching.

**Comment by sjrd** [comment]
That's all.

**Comment by mrmiywj** [comment]
Changes:
- Correct code smells (as I thought)
- Some exception test cases

File: build.sbt [hunk 28]
```diff
@@ -12,7 +12,7 @@ libraryDependencies ++= Seq(
   "org.scala-js" %% "scalajs-tools" % scalaJSVersion,
   "org.scala-js" %% "scalajs-js-envs" % scalaJSVersion,
   "com.lihaoyi" %% "fastparse" % "0.3.7",
-  
+  "org.scalatest" % "scalatest_2.11" % "2.2.6",
```
Comment (sjrd) [inline build.sbt]: Add `% "test"`. This is a test-only dependency.

File: src/main/scala/calc/Compiler.scala [hunk 29]
```diff
@@ -61,21 +62,107 @@ object Compiler {
...
+                                                    => env+(s.name -> irtpe.DoubleType))
+        val unboxParams = idList.map(s =>
+          irt.VarDef(irt.Ident(s.name), irtpe.DoubleType, mutable = false,
+            irt.Unbox(irt.VarRef(irt.Ident("$$"+s.name))(irtpe.AnyType), 'D')))
+        val bl = irt.Block(unboxParams ++ List(compileExpr(body, newTypeEnv)))
+        irt.Closure(captureParam, params, bl, captureList.map(s => irt.VarRef(irt.Ident(s))(typeEnv(s))))
+
+      case Call(Ident(name), args) =>
+        try {
+          if (functionTypes.contains(name)) {
```
Comment (sjrd) [inline src/main/scala/calc/Compiler.scala]: In case `name` is defined _both_ in `functionTypes` and in `typeEnv`, the one from `typeEnv` should take precedence, so that shadowing works: `let sin = 5 in sin` should work and yield 5, as well as `let sin = fun(a, b) = { a + b } in sin(3, 2)`.

**Comment by sjrd** [comment]
That's all.

**Comment by mrmiywj** [comment]
Changes:
Correct the precedence and add some tests.

Thanks for your replies and you're so kind:)

**Comment by sjrd** [comment]
> Changes: Correct the precedence and add some tests.

I think you forgot to push ;)

**Comment by mrmiywj** [comment]
Sorry for this =)

**Comment by sjrd** [comment]
OK, it looks pretty complete, now. Well done! I hope this assignment was helpful in discovering the IR a bit ;)

**Comment by mrmiywj** [comment]
Aha! It's done!
Besides discovering IR, you also help me correct some bad coding methodologies. Thank you!

**Comment by sjrd** [comment]
Closing since this is finished.
